modules/notification/notification.go

```diff
@@ -1,50 +1,168 @@
-// Copyright 2016 The Gitea Authors. All rights reserved.
+// Copyright 2018 The Gitea Authors. All rights reserved.
 // Use of this source code is governed by a MIT-style
 // license that can be found in the LICENSE file.
 
 package notification
 
 import (
+	"code.gitea.io/git"
 	"code.gitea.io/gitea/models"
-	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/notification/base"
+	"code.gitea.io/gitea/modules/notification/ui"
 )
 
-type (
-	notificationService struct {
-		issueQueue chan issueNotificationOpts
+var (
+	notifiers []base.Notifier
+)
+
+// RegisterNotifier providers method to receive notify messages
+func RegisterNotifier(notifier base.Notifier) {
+	go notifier.Run()
+	notifiers = append(notifiers, notifier)
+}
+
+func init() {
+	RegisterNotifier(ui.NewNotifier())
+}
+
+// NotifyCreateIssueComment notifies issue comment related message to notifiers
+func NotifyCreateIssueComment(doer *models.User, repo *models.Repository,
+	issue *models.Issue, comment *models.Comment) {
+	for _, notifier := range notifiers {
+		notifier.NotifyCreateIssueComment(doer, repo, issue, comment)
 	}
+}
 
-	issueNotificationOpts struct {
-		issue                *models.Issue
-		notificationAuthorID int64
+// NotifyNewIssue notifies new issue to notifiers
+func NotifyNewIssue(issue *models.Issue) {
+	for _, notifier := range notifiers {
+		notifier.NotifyNewIssue(issue)
 	}
-)
+}
 
-var (
-	// Service is the notification service
-	Service = &notificationService{
-		issueQueue: make(chan issueNotificationOpts, 100),
+// NotifyIssueChangeStatus notifies close or reopen issue to notifiers
+func NotifyIssueChangeStatus(doer *models.User, issue *models.Issue, closeOrReopen bool) {
+	for _, notifier := range notifiers {
+		notifier.NotifyIssueChangeStatus(doer, issue, closeOrReopen)
 	}
-)
+}
 
-func init() {
-	go Service.Run()
+// NotifyMergePullRequest notifies merge pull request to notifiers
+func NotifyMergePullRequest(pr *models.PullRequest, doer *models.User, baseGitRepo *git.Repository) {
+	for _, notifier := range notifiers {
+		notifier.NotifyMergePullRequest(pr, doer, baseGitRepo)
```

Comment:
What do you think about running all notifiers in their own goroutine?

Reply:
@JonasFranzDEV Every notifier could create worker go routines via channels like what ui and mail did currently. Otherwise, every request will create many go routines that I don't think it's a good idea.

Reply:
@lunny agree